### ISSUE TITLE
Analyze parameter type for assign operation SQL

### ISSUE THREAD

**Issue by silathdiir** [alex-dukhno/database#400]
Some database drivers send Parse message of Extended Protocol with parameter type of OID value `zero` which we could not handle now. The parameter type of OID zero means unspecified type. We needs to analyze the real parameter type from the SQL statement.

This issue could be reproduced by compatibility tests in https://github.com/alex-dukhno/database/blob/master/tests/compatibility/src/test/groovy/io/database/extended/SelectiveAttributesQueryOperationsSpec.groovy. We added `@ignore` for most test cases. You could run compatibility tests via command `local/compatibility.sh` or `gradle test --tests io.database.extended.SelectiveAttributesQueryOperationsSpec`.

After some investigation, we may reference the materialize code in https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan.rs#L289. The `describe` function has a parameter of type `&[Option<pgrepr::Type>]`, `Option::None` means unspecified type, and returns the real parameter types (`Vec<pgrepr::Type>`).

According to the discussion in PR #397 , it seems that the code should be implemented in `query_analyzer`.

For this issue we only handle assign operations as below (may add a new file named `assign_operation.rs` in `query_analyzer`.

```
insert into <table> values (?, ?, ?)
```

**Comment by silathdiir** [comment]
Hi @alex-dukhno , please assign this issue to me. Thanks.

**Comment by alex-dukhno** [comment]
@silathdiir, I am looking in this one too. It appears a blocker for other tasks for parameter analysis

**Comment by silathdiir** [comment]
Hi @alex-dukhno , sorry for late reply. Please feel free to work on this one. I am sorry that I was blocked on other thing. I only started on compatibility test case and passed in `mut &[PgType]` to the `describe` function of `query_analyzer`.
I will unsign me and try to investigate #194 from next week if no one was looking into it. Thanks.

**Comment by alex-dukhno** [comment]
> started on compatibility test case and passed in `mut &[PgType]` to the `describe` function of `query_analyzer`.

Can you explain you thought on this? Maybe you give me a hint ... I started debugging PG JDBC driver 😄 and seems like we aren't respecting extended query protocol flow, at least from the driver perspective

**Comment by silathdiir** [comment]
Hi @alex-dukhno , I think this issue is caused by the `groovy` database driver passing OID zero for all parameter types. There isn't a parameter could be used to pass the parameter types.

Running the simple test case (like our compatibility tests) as below:

```
pg.executeUpdate 'insert into SCHEMA_NAME.TABLE_NAME values (?, ?, ?)', [1, 2, 3]
```

I am sorry that I missed to handle SQL analyzing for parameter type of OLD zero. Since I implemented the Extended Protocol testing with [epgsql](https://github.com/epgsql/epgsql), running the test case as below:

```
{ok, S} = epgsql:parse(C, "insert into schema_name.table_name values ($1, $2, $3)", [int4, int4, int4]).
```

I think it works since the parameter type `int4` (`small_int` OID-21) is passed to function `epgsql:parse`. So the right parameter type `small_int` could be saved to `prepared_statement` without analyzing the original SQL statement.

After reviewing the `materialize` [code](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan.rs#L289), I supposed to not handle OLD zero with `None` (as `materialize` implementation), but added `Unspecified` to PgType as below.

```
pub enum PgType {
  ...
+    /// Unspecified type maps OID value of zero. It needs to be analyzed to a
+    /// valid data type before using.
+    Unspecified,
 }
```

And passing `&mut [PgType]` to `describe` function in `src/query_analyzer/src/lib.rs` as blow.

```
-    pub fn describe(&self, statement: &Statement) -> Result<Description, DescriptionError> {
+    pub fn describe(&self, statement: &Statement, param_types: &mut [PgType]) -> Result<Description, DescriptionError> {
```

So we could inspect and update the real parameter types by SQL analysis. I think the real parameter types could be parsed via the value types of table columns (or `UnaryOp` and `BinaryOp`) by regression calls for each `Exprs`.

If you are not hurry, please assign this issue to me again, I think I could try to finish it in two weeks. Sorry for delay.

**Comment by alex-dukhno** [comment]
I went ahead and published pg_wire with new conversion function https://github.com/alex-dukhno/pg_wire/blob/main/src/types.rs#L71 also noticed during debugging that database doesn't handle `DescribePortal` message correctly. Right now, I am trying to figure out why we send `ParameterDescription` that is not needed

**Comment by alex-dukhno** [comment]
@silathdiir I've figured out last bit of the problem. We sent `ReadyForQuery` message after `Execute` and `Sync` messages from the driver during extended query protocol flow - I fixed it in https://github.com/alex-dukhno/database/pull/407
However, I think it is ad-hoc solution and I'll be working on state machine of message exchanging between client and server.

**Comment by alex-dukhno** [comment]
I have a look into `query_analizer` code and I don't know if this task should be closed ...
column types are handled basically here https://github.com/alex-dukhno/database/blob/master/src/query_analyzer/src/lib.rs#L36
🤷

**Comment by silathdiir** [comment]
Hi @alex-dukhno , for `ParameterDescription`, it seems that it should be handled in `Parse` process. It could be found in below places:

PostgreSQL doc: https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY (search `ParameterDescription`)

`materialize` source code: https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/protocol.rs#L517

`epsql` (`erlang` database driver) source code: https://github.com/epgsql/epgsql/blob/devel/src/commands/epgsql_cmd_parse.erl#L12

**Comment by silathdiir** [comment]
I think you have already fixed this issue for `insert` compatibility test in #407 . We could try to fix the ignored test and more complicated ones in another issue. Thanks.

**Comment by alex-dukhno** [comment]
We don't handle situation when number of parameter != table columns.
I've ignored test https://github.com/alex-dukhno/database/blob/master/src/node/src/query_engine/tests/simple_prepared_statement.rs#L23, I am sure same situation with extended query flow

**Comment by silathdiir** [comment]
Hi @alex-dukhno , please assign back to me. I saw you added a TODO comment. I'd like to work on this. Thanks.
